### PR TITLE
feat(mobile-client): hosted-ui auth response handler is now built into redirect activity

### PR DESCRIPTION
**Pull request by raphkim**
*Issue #, if available:*
#2291 

*Description of changes:*
Fixes `signOut` after `showSignIn()` from failing to clear browser cookies.

## Issue
Previously, `AuthClient#signOut()` would do the following:
1. clear cache (marks current state as "signed out")
2. launch sign out redirect URL for clearing browser cookies

Step 2 was non-blocking, which meant that it could be cancelled by switching activity, resulting in a state where cookies are _not_ cleared, but the client mistakenly believes that it is currently signed out. This results in the next sign-in attempt not prompting for the credentials.

## Fix
Auth redirect handler is now added to `HostedUIRedirectActivity#onResume()` (called upon finishing the flow regardless of whether it was successful or canceled), which will then trigger the callbacks to let `AWSMobileClient` know that auth flow was finished. This will let `AWSMobileClient` know that sign-in or sign-out was either successful or canceled midway and handle the result appropriately.

There is an added bonus with this PR that now the customer no longer needs to call the `AWSMobileClient#handleAuthResponse(Intent)` method in their own app for handling auth results.

## TLDR
Hosted UI sign-out will now work _**without any breaking change**_, but the fix will only be applied if, `AndroidManifest.xml` file is updated to use the new `HostedUIRedirectActivity`.

## TODO
- ~~Add official documentation for v2 SDK~~ https://github.com/aws-amplify/docs/pull/3214
- Add official documentation for Amplify

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws-android-sdk-mobile-client/src/main/java/com/amazonaws/mobile/client/activities/CustomTabsRedirectActivity.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import android.app.Activity;
+import android.os.Bundle;
+
+import com.amazonaws.mobile.client.AWSMobileClient;
+import com.amazonaws.mobileconnectors.cognitoauth.activities.CustomTabsManagerActivity;
+
+/**
+ * Handles auth redirect for sign-in and sign-out.
+ */
+public final class CustomTabsRedirectActivity extends Activity {
```
[the remark this answers is not in the document]
Comment (raphkim) [reply aws-android-sdk-mobile-client/src/main/java/com/amazonaws/mobile/client/activities/CustomTabsRedirectActivity.java:12]: Yep, the customer won't be using this file anywhere besides in the manifest. I think it would be better to keep the same name to avoid confusion if anything.

**Review by rjuliano — APPROVED** [review]
Good sleuthing on this one. Ship it 🚢

File: aws-android-sdk-cognitoauth/src/main/java/com/amazonaws/mobileconnectors/cognitoauth/AuthClient.java [hunk 2]
```diff
@@ -271,18 +279,33 @@ public void signOut(final boolean clearLocalTokensOnly) {
...
     public void signOut(final boolean clearLocalTokensOnly, final String browserPackage) {
-        LocalDataManager.clearCache(pool.awsKeyValueStore, context, pool.getAppId(), userId);
+        // Try deleting cookies.
         if (!clearLocalTokensOnly) {
+            endSession(browserPackage);
+        }
+
+        // Delete local cache
+        LocalDataManager.clearCache(pool.awsKeyValueStore, context, pool.getAppId(), userId);
+    }
```
[the remark this answers is not in the document]
Comment (raphkim) [reply aws-android-sdk-cognitoauth/src/main/java/com/amazonaws/mobileconnectors/cognitoauth/AuthClient.java]: this logic is now outdated. `endSession` will result in an exception being thrown if redirect fails.

File: aws-android-sdk-mobile-client/src/main/java/com/amazonaws/mobile/client/activities/CustomTabsRedirectActivity.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+        startActivity(CustomTabsManagerActivity.createResponseHandlingIntent(
+                this, getIntent().getData()));
+    }
+
+    @Override
+    public void onResume() {
+        super.onResume();
+        AWSMobileClient.getInstance().handleAuthResponse(getIntent());
+        finish();
+    }
```
[the remark this answers is not in the document]
Comment (raphkim) [reply aws-android-sdk-mobile-client/src/main/java/com/amazonaws/mobile/client/activities/CustomTabsRedirectActivity.java]: renamed to `HostedUIRedirectActivity`

File: aws-android-sdk-cognitoauth/src/main/java/com/amazonaws/mobileconnectors/cognitoauth/AuthClient.java [hunk 4]
```diff
@@ -708,4 +768,36 @@ public void onServiceDisconnected(final ComponentName name) {
...
+                        return true;
+                    }
+                }
+            }
+            Log.w(TAG, REDIRECT_ACTIVITY_NAME + " is not declared in AndroidManifest.");
+        } catch (Exception error) {
+            Log.w(TAG, "Failed to inspect packages.");
+        }
+        return false;
+    }
```
Comment (raphkim) [inline aws-android-sdk-cognitoauth/src/main/java/com/amazonaws/mobileconnectors/cognitoauth/AuthClient.java:802]: This is to make sure that apps that:
- use the updated library
- did not update the code to use new `HostedUIRedirectActivity`
 
will not be held hostage by the signout-wait-latch that will never count down due to lack of callback event. We will inspect their android manifest to make sure that they followed the newest documentation to use `HostedUIRedirectActivity` before enabling the latch that fixes the race condition.

File: aws-android-sdk-cognitoauth/src/main/java/com/amazonaws/mobileconnectors/cognitoauth/AuthClient.java [hunk 5]
```diff
@@ -271,18 +302,42 @@ public void signOut(final boolean clearLocalTokensOnly) {
...
             launchSignOut(pool.getSignOutRedirectUri(), browserPackage);
+            if (!isRedirectActivityDeclared()) {
+                cookiesCleared.countDown();
+            }
+            redirectReceived = cookiesCleared.await(REDIRECT_TIMEOUT_SECONDS, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
+            throw new AuthNavigationException("User cancelled sign-out.");
+        }
+        if (!redirectReceived) {
+            throw new AuthServiceException("Timed out while waiting for sign-out redirect response.");
```
Comment (raphkim) [inline aws-android-sdk-cognitoauth/src/main/java/com/amazonaws/mobileconnectors/cognitoauth/AuthClient.java:340]: These are runtime exceptions so they do not introduce any breaking change.

Neither exception will ever be thrown if they are not using the new `HostedUIRedirectActivity` either.

**Review by rjuliano — APPROVED** [review]
LGTM 🚢

**Review by richardmcclellan — APPROVED** [review]
This is great @raphkim!  I created a sample app, following your updated documentation, and the sign out behavior is working perfectly now, even if I call `startActivity` from the sign out success callback.  Great work! 🚀
